<!DOCTYPE html>
<html lang="en"><head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1"><!-- Begin Jekyll SEO tag v2.6.1 -->
<title>Installation | MIGNON</title>
<meta name="generator" content="Jekyll v3.9.0" />
<meta property="og:title" content="Installation" />
<meta name="author" content="Martin Garrido Rodriguez-Cordoba" />
<meta property="og:locale" content="en_US" />
<meta name="description" content="Bioinformatic workflow for the mechanistic integrative analysis of RNA-Seq data" />
<meta property="og:description" content="Bioinformatic workflow for the mechanistic integrative analysis of RNA-Seq data" />
<link rel="canonical" href="http://localhost:4000/1_installation.html" />
<meta property="og:url" content="http://localhost:4000/1_installation.html" />
<meta property="og:site_name" content="MIGNON" />
<script type="application/ld+json">
{"@type":"WebPage","headline":"Installation","url":"http://localhost:4000/1_installation.html","author":{"@type":"Person","name":"Martin Garrido Rodriguez-Cordoba"},"description":"Bioinformatic workflow for the mechanistic integrative analysis of RNA-Seq data","@context":"https://schema.org"}</script>
<!-- End Jekyll SEO tag -->
<link rel="stylesheet" href="/assets/main.css"><link type="application/atom+xml" rel="alternate" href="http://localhost:4000/feed.xml" title="MIGNON" /></head>
<body><header class="site-header" role="banner">

  <div class="wrapper"><a class="site-title" rel="author" href="/">MIGNON</a><nav class="site-nav">
        <input type="checkbox" id="nav-trigger" class="nav-trigger" />
        <label for="nav-trigger">
          <span class="menu-icon">
            <svg viewBox="0 0 18 15" width="18px" height="15px">
              <path d="M18,1.484c0,0.82-0.665,1.484-1.484,1.484H1.484C0.665,2.969,0,2.304,0,1.484l0,0C0,0.665,0.665,0,1.484,0 h15.032C17.335,0,18,0.665,18,1.484L18,1.484z M18,7.516C18,8.335,17.335,9,16.516,9H1.484C0.665,9,0,8.335,0,7.516l0,0 c0-0.82,0.665-1.484,1.484-1.484h15.032C17.335,6.031,18,6.696,18,7.516L18,7.516z M18,13.516C18,14.335,17.335,15,16.516,15H1.484 C0.665,15,0,14.335,0,13.516l0,0c0-0.82,0.665-1.483,1.484-1.483h15.032C17.335,12.031,18,12.695,18,13.516L18,13.516z"/>
            </svg>
          </span>
        </label>

        <div class="trigger"><a class="page-link" href="/1_installation.html">Installation</a><a class="page-link" href="/2_input.html">Input</a><a class="page-link" href="/3_output.html">Output</a><a class="page-link" href="/4_advanced.html">Advanced</a></div>
      </nav></div>
</header>
<main class="page-content" aria-label="Content">
      <div class="wrapper">
        <h1 id="quick-start">Quick start</h1>

<p>Users with experience on workflow managers and containers can directly clone and execute the dry run of the workflow if they already have <a href="https://git-scm.com/">Git</a>, <a href="https://www.java.com/es/download/">Java (v1.8.0)</a> and <a href="https://www.docker.com/">docker</a> installed. This can be done by executing the following commands on your terminal:</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>git clone https://github.com/babelomics/MIGNON.git
cd MIGNON
bash runMignonExample.sh
</code></pre></div></div>

<p>After completing the dry runs, the results of the pipeline can be found at the <code class="language-plaintext highlighter-rouge">dry_run/cromwell-executions</code> directory.</p>

<p>On the other hand, we encourage users with less experience to fully read the installation guide before executing the workflow.</p>

<h1 id="requirements">Requirements</h1>

<p>From a programmatic perspective, MIGNON is a chain of tasks written in the <a href="https://github.com/openwdl/wdl">Workflow Description Language</a>. Users need to install and download the following software to be able to run it:</p>

<h2 id="java"><a href="https://www.java.com/es/download/">Java</a></h2>

<p>Java is used to execute the workflow management software that interprets and launch the different tasks within the pipeline. Once installed, users can check the version with <code class="language-plaintext highlighter-rouge">java -version</code>:</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>$ java -version
openjdk version "1.8.0_275"
OpenJDK Runtime Environment (build 1.8.0_275-8u275-b01-0ubuntu1~18.04-b01)
OpenJDK 64-Bit Server VM (build 25.275-b01, mixed mode)
</code></pre></div></div>

<h2 id="docker"><a href="https://www.docker.com/">Docker</a></h2>

<p>Docker (or any engine able to run docker containers, as <a href="https://sylabs.io/docs/">Singularity</a> is used to execute all the tasks of the workflow within an isolated unit of containerized software, freeing users to install each of the needed components of the pipeline. Once installed, users should be able to get the following output after executing <code class="language-plaintext highlighter-rouge">docker run hello-world</code>:</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>$ docker run hello-world

Hello from Docker!
This message shows that your installation appears to be working correctly.

To generate this message, Docker took the following steps:
 1. The Docker client contacted the Docker daemon.
 2. The Docker daemon pulled the "hello-world" image from the Docker Hub.
    (amd64)
 3. The Docker daemon created a new container from that image which runs the
    executable that produces the output you are currently reading.
 4. The Docker daemon streamed that output to the Docker client, which sent it
    to your terminal.

To try something more ambitious, you can run an Ubuntu container with:
 $ docker run -it ubuntu bash

Share images, automate workflows, and more with a free Docker ID:
 https://hub.docker.com/

For more examples and ideas, visit:
 https://docs.docker.com/get-started/
</code></pre></div></div>

<h2 id="cromwell"><a href="https://github.com/broadinstitute/cromwell">Cromwell</a></h2>

<p>Users need to download the <strong>cromwell-XX.jar</strong> file that can be retrieved from the cromwell releases page. Particularly, MIGNON was tested with the release 47 of cromwell, which can be obtained in the <a href="https://github.com/broadinstitute/cromwell/releases/tag/47">following link</a>. Cromwell is the scientific workflow engine that interprets and executes the pipeline. It is the software why users need to have Java v1.8.0 installed. <strong>Note</strong>: When executing the <code class="language-plaintext highlighter-rouge">runMignonExample.sh</code> script, a copy of the cromwell binaries is downloaded together with the example data.</p>

<h2 id="mignon"><a href="https://github.com/babelomics/MIGNON/">MIGNON</a></h2>

<p>Users can download the latest MIGNON code from our Github repository. Users with <a href="https://git-scm.com/">git</a> can clone the repository using:</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>$ git clone https://github.com/babelomics/MIGNON.git
</code></pre></div></div>
<p>On the other hand, if you are not using git, you can get a zipped version of the code on the following link:</p>

<ul>
  <li><a href="https://github.com/babelomics/MIGNON/archive/master.zip">—-&gt; <strong>Link to the compressed version of MIGNON</strong> &lt;—-</a>.</li>
</ul>

<h1 id="run-the-workflow">Run the workflow</h1>

<h2 id="dry-run">Dry run</h2>

<p>After fulfilling all the requirements, users can test the workflow by performing a <em>dry run</em> with example data. This can be done by calling the <code class="language-plaintext highlighter-rouge">runMignonExample.sh</code> script that is located in the root directory of the repo. To do so, please execute the following command within the MIGNON folder:</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>$ bash runMignonExample.sh
</code></pre></div></div>
<p>This script will:</p>

<ol>
  <li>Download all the required data to perform the dry run into the mignon_test_data folder. These data includes:
    <ul>
      <li>A reduced version of the human reference genome, transcriptome and annotations obtained from <a href="https://www.ensembl.org/Homo_sapiens/Info/Index">ENSEMBL</a>.</li>
      <li>A subset of the Variant Effect Predictor cache file obtained from <a href="ftp://ftp.ensembl.org/pub/release-99/variation/indexed_vep_cache/">ENSEMBL</a>.</li>
      <li>A reduced version of the dbSNP database in VCF format obtained from the <a href="https://ftp.ncbi.nih.gov/snp/organisms/human_9606_b150_GRCh38p7/VCF/All_20170710.vcf.gz">NCBI</a>.</li>
      <li>The HISAT2, STAR and Salmon indexes created from the reduced version of the reference genome and transcriptome.</li>
      <li>Four reduced paired read files that were obtained after aligning the whole samples to the reference genome and extract the reads that were found in a particular region of the chromosome 9.</li>
      <li>A copy of the cromwell and womtool binaries. The cromwell license that allow the re-distribution of such binaries can be found in the <code class="language-plaintext highlighter-rouge">LICENSE-cromwell</code> file in the root directory of MIGNON.</li>
    </ul>
  </li>
  <li>
    <p>Execute a dry run of the workflow, where the different execution modes will be tested. Please take into account that, as MIGNON will execute all the different steps included in the workflow under different configurations, this dry run may take a while.</p>
  </li>
  <li>Once finished, it will print a final message indicating the success of the dry run and the ability of the user to execute MIGNON!</li>
</ol>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>MIGNON: Success!! Dry run completed. You are ready to execute MIGNON.
</code></pre></div></div>

<h2 id="regular-run">Regular run</h2>

<p>Although the dry run is very useful to test that all the MIGNON dependencies were installed successfully, users need to prepare their inputs to perform a regular run of the pipeline. To guide users through the different inputs required by MIGNON, we have prepared the <a href="/2_input.html">input section</a>. Once with the input prepared in the JSON format, users can run the workflow by executing the following command in a terminal:</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>$ java -Dconfig.file=/path/to/config_file.conf -jar /path/to/cromwell.jar run /path/to/MIGNON.wdl -i /path/to/input.json
</code></pre></div></div>

<h2 id="troubleshooting">Troubleshooting</h2>

<p>If you have any doubt about MIGNON features, inputs and outputs, please contact us on the <a href="https://github.com/babelomics/MIGNON/issues">issues</a> section! On the other hand, if you have trouble installing or executing Java, Docker or Cromwell, we encourage users to use their respective forums to address their questions. Finally, we strongly recommend to read the <a href="https://cromwell.readthedocs.io/en/stable/tutorials/FiveMinuteIntro/">5 minutes introduction</a> to cromwell written by the developers of the tool.</p>

      </div>
    </main><footer class="site-footer h-card">
  <data class="u-url" href="/"></data>

  <div class="wrapper">

    <h2 class="footer-heading">MIGNON</h2>

    <div class="footer-col-wrapper">
      <div class="footer-col footer-col-1">
        <ul class="contact-list">
          <li class="p-name">Martin Garrido Rodriguez-Cordoba</li></ul>
      </div>

      <div class="footer-col footer-col-2"><ul class="social-media-list"><li><a href="https://github.com/babelomics"><svg class="svg-icon"><use xlink:href="/assets/minima-social-icons.svg#github"></use></svg> <span class="username">babelomics</span></a></li><li><a href="https://www.twitter.com/ClinicalBioinfo"><svg class="svg-icon"><use xlink:href="/assets/minima-social-icons.svg#twitter"></use></svg> <span class="username">ClinicalBioinfo</span></a></li></ul>
</div>

      <div class="footer-col footer-col-3">
        <p>Bioinformatic workflow for the mechanistic integrative analysis of RNA-Seq data</p>
      </div>
    </div>

  </div>

</footer>
</body>

</html>
