<!DOCTYPE html>
<html lang="en"><head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1"><!-- Begin Jekyll SEO tag v2.6.1 -->
<title>Input | MIGNON</title>
<meta name="generator" content="Jekyll v3.9.0" />
<meta property="og:title" content="Input" />
<meta name="author" content="Martin Garrido Rodriguez-Cordoba" />
<meta property="og:locale" content="en_US" />
<meta name="description" content="Bioinformatic workflow for the mechanistic integrative analysis of RNA-Seq data" />
<meta property="og:description" content="Bioinformatic workflow for the mechanistic integrative analysis of RNA-Seq data" />
<link rel="canonical" href="http://localhost:4000/2_input.html" />
<meta property="og:url" content="http://localhost:4000/2_input.html" />
<meta property="og:site_name" content="MIGNON" />
<script type="application/ld+json">
{"@type":"WebPage","headline":"Input","url":"http://localhost:4000/2_input.html","author":{"@type":"Person","name":"Martin Garrido Rodriguez-Cordoba"},"description":"Bioinformatic workflow for the mechanistic integrative analysis of RNA-Seq data","@context":"https://schema.org"}</script>
<!-- End Jekyll SEO tag -->
<link rel="stylesheet" href="/assets/main.css"><link type="application/atom+xml" rel="alternate" href="http://localhost:4000/feed.xml" title="MIGNON" /></head>
<body><header class="site-header" role="banner">

  <div class="wrapper"><a class="site-title" rel="author" href="/">MIGNON</a><nav class="site-nav">
        <input type="checkbox" id="nav-trigger" class="nav-trigger" />
        <label for="nav-trigger">
          <span class="menu-icon">
            <svg viewBox="0 0 18 15" width="18px" height="15px">
              <path d="M18,1.484c0,0.82-0.665,1.484-1.484,1.484H1.484C0.665,2.969,0,2.304,0,1.484l0,0C0,0.665,0.665,0,1.484,0 h15.032C17.335,0,18,0.665,18,1.484L18,1.484z M18,7.516C18,8.335,17.335,9,16.516,9H1.484C0.665,9,0,8.335,0,7.516l0,0 c0-0.82,0.665-1.484,1.484-1.484h15.032C17.335,6.031,18,6.696,18,7.516L18,7.516z M18,13.516C18,14.335,17.335,15,16.516,15H1.484 C0.665,15,0,14.335,0,13.516l0,0c0-0.82,0.665-1.483,1.484-1.483h15.032C17.335,12.031,18,12.695,18,13.516L18,13.516z"/>
            </svg>
          </span>
        </label>

        <div class="trigger"><a class="page-link" href="/1_installation.html">Installation</a><a class="page-link" href="/2_input.html">Input</a><a class="page-link" href="/3_output.html">Output</a><a class="page-link" href="/4_advanced.html">Advanced</a></div>
      </nav></div>
</header>
<main class="page-content" aria-label="Content">
      <div class="wrapper">
        <h1 id="main-inputs">Main inputs</h1>

<h2 id="json">JSON</h2>

<p>As explained by <a href="https://github.com/openwdl/wdl/blob/master/versions/development/SPEC.md#specifying-workflow-inputs-in-json">the Workflow Description Language (WDL) authors</a>, cromwell uses a <a href="https://www.json.org/">JSON</a> formatted file as main input. In MIGNON, this file contains <strong>absolute</strong> paths to the files that are used during the analysis, <strong>decision variables</strong> to control which workflow steps are carried out (e.g. wether to perform or not the variant calling) and <strong>arguments</strong> to control each task execution (e.g. the amount of threads to use for the alignment).</p>

<p>Here you can find an example the content of a MIGNON input JSON file. To build your own input file, please go to the <a href="#preparing-the-json"><strong>preparing the JSON</strong></a> section.</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>{
    "MIGNON.execution_mode": "hisat2",
    "MIGNON.is_paired_end": true,
    "MIGNON.do_vc": true,
	"MIGNON.input_fastq_r1": [
		"flagForPwd/mignon_test_data/subset_fastq/SRR8615222_1.fastq.gz",
		"flagForPwd/mignon_test_data/subset_fastq/SRR8615223_1.fastq.gz",
		"flagForPwd/mignon_test_data/subset_fastq/SRR8615224_1.fastq.gz",
		"flagForPwd/mignon_test_data/subset_fastq/SRR8615225_1.fastq.gz"
	],
    "MIGNON.input_fastq_r2": [
		"flagForPwd/mignon_test_data/subset_fastq/SRR8615222_2.fastq.gz",
		"flagForPwd/mignon_test_data/subset_fastq/SRR8615223_2.fastq.gz",
		"flagForPwd/mignon_test_data/subset_fastq/SRR8615224_2.fastq.gz",
		"flagForPwd/mignon_test_data/subset_fastq/SRR8615225_2.fastq.gz"
	],
    "MIGNON.sample_id": [
        "Control_1",
        "Control_2",
        "Problem_1",
        "Problem_2"
    ],
    "MIGNON.group": [
        "Control",
        "Control",
        "Problem",
        "Problem"
    ],
    "MIGNON.gtf_file": "flagForPwd/mignon_test_data/Homo_sapiens.GRCh38.99.chr.gtf",
    "MIGNON.hisat2_index_path": "flagForPwd/mignon_test_data/hisat_index",
    "MIGNON.hisat2_index_prefix": "genome",
    "MIGNON.vep_cache_dir": "flagForPwd/mignon_test_data/subset_vep_cache",
    "MIGNON.ref_fasta": "flagForPwd/mignon_test_data/subset_genome.fa",
    "MIGNON.ref_fasta_index": "flagForPwd/mignon_test_data/subset_genome.fa.fai",
    "MIGNON.ref_dict": "flagForPwd/mignon_test_data/subset_genome.dict",
    "MIGNON.db_snp_vcf": "flagForPwd/mignon_test_data/subset_variants.vcf.gz",
    "MIGNON.db_snp_vcf_index": "flagForPwd/mignon_test_data/subset_variants.vcf.gz.tbi",
    "MIGNON.known_vcfs":["flagForPwd/mignon_test_data/subset_variants.vcf.gz"],
    "MIGNON.known_vcfs_indices": ["flagForPwd/mignon_test_data/subset_variants.vcf.gz.tbi"],
    "MIGNON.edger_script": "flagForPwd/scripts/edgeR.r",
    "MIGNON.hipathia_script": "flagForPwd/scripts/hipathia.r"

}
</code></pre></div></div>

<h2 id="cromwell-configuration-file">Cromwell configuration file</h2>

<p>The configuration file tells cromwell how to handle and launch the jobs that make up the workflow. When executing a single job, it places the values from the WDL file in the appropiated position within the container engine command. For example, the <code class="language-plaintext highlighter-rouge">requested_memory</code> runtime input, which controls the memory that is allocated for the execution of the task, is passed to the <code class="language-plaintext highlighter-rouge">--memory</code> argument when using <code class="language-plaintext highlighter-rouge">docker run</code>. On the other hand, when executing the workflow on a HPC environment, the <code class="language-plaintext highlighter-rouge">requested_memory</code> runime input is used in the submission of the job with <code class="language-plaintext highlighter-rouge">sbatch --mem</code> (SLURM command). In addition, this file can control other top-level features of cromwell, as for example how the duplicated files are managed or <strong>the number of parallel jobs that can be executed</strong>. You can find more information and examples of the possible backends that can be configured through this file in the <a href="https://cromwell.readthedocs.io/en/stable/Configuring/">cromwell documentation</a>.</p>

<p>To run MIGNON, we have prepared two different configurations files. The first one, <a href="https://github.com/babelomics/MIGNON/blob/master/configs/LocalWithDocker.conf"><code class="language-plaintext highlighter-rouge">LocalWithDocker.conf</code></a>, is prepared to run MIGNON locally using docker to execute the containerized software. The second one, <a href="https://github.com/babelomics/MIGNON/blob/master/configs/SlurmAndSingularity.conf"><code class="language-plaintext highlighter-rouge">SlurmAndSingularity.conf</code></a>, is intended to be used within HPC environments that use <a href="https://slurm.schedmd.com/documentation.html">Slurm</a> as workload manager and have <a href="https://sylabs.io/docs/">Singularity</a> installed as a module to run the containerized software.</p>

<h1 id="execution-modes">Execution modes</h1>

<p>There are 5 different execution modes of MIGNON. Each of them makes use of a different combination of tools in the intial steps of the workflow. For example, the “salmon-hisat2” execution mode uses <a href="http://www.ccb.jhu.edu/software/hisat/index.shtml">HISAT2</a> to perform the alignment and <a href="https://combine-lab.github.io/salmon/">Salmon</a> to perform the gene expression quantification from reads. The HISAT2 alignments can then be used to run the variant calling (VC) sub-workflow with <a href="https://gatk.broadinstitute.org/hc/en-us">GATK</a> to obtain the genomic information. Each mode has a specific computational profile and it should be selected depending on the user needs. In a nutshell, on modes where [STAR] (https://github.com/alexdobin/STAR) is used, the maximum memory requirements and velocity of the alignment are high. On the other hand, on modes where <a href="http://www.ccb.jhu.edu/software/hisat/index.shtml">HISAT2</a> is used the memory requirements and velocity of alignments are lower, allowing more parallel jobs for the same computational price. The following table summarizes the basic properties of each execution mode:</p>

<table>
  <thead>
    <tr>
      <th>Execution mode</th>
      <th>Alignment</th>
      <th>Quantification</th>
      <th>Allows VC</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>“salmon-hisat2”</td>
      <td>HISAT2</td>
      <td>Salmon</td>
      <td>Yes</td>
    </tr>
    <tr>
      <td>“salmon-star”</td>
      <td>STAR</td>
      <td>Salmon</td>
      <td>Yes</td>
    </tr>
    <tr>
      <td>“hisat2”</td>
      <td>HISAT2</td>
      <td>featureCounts</td>
      <td>Yes</td>
    </tr>
    <tr>
      <td>“star”</td>
      <td>STAR</td>
      <td>featureCounts</td>
      <td>Yes</td>
    </tr>
    <tr>
      <td>“salmon”</td>
      <td>-</td>
      <td>Salmon</td>
      <td>No</td>
    </tr>
  </tbody>
</table>

<p>In addition, the following figure shows which steps of the workflow are carried out under each execution mode:</p>

<p><img src="/img/MIGNON_modes.svg" alt="execution_modes" /></p>

<p><strong>Depending on the execution modes, the required values in the input JSON can change</strong>. We strongly recommend to use the combined execution modes “salmon-star” or “salmon-hisat2”, as they use the pseudo-alignment strategy to quantify gene expression dealing with the multi-mapping reads problem, and star or hisat2 to obtain the alignments for the variant calling sub-workflow.</p>

<h1 id="preparing-the-json">Preparing the JSON</h1>

<p>As explained before, the JSON file is used to store and register the file paths, workflow variables and execution parameters. For MIGNON, the values that should be included in this file can be divided in two different types: <a href="#required-values">Required values</a>, <a href="#default-values">default values</a>.</p>

<h2 id="required-values">Required values</h2>

<p>The values detailed in this section are mandatory and will vary depending on the execution mode. Apart from the input reads or sample ids, it is important to pay attention to the different reference material that is required to perform the alignment, pseudo-alignment or variant calling. The following table contains the basic information about each value, the execution modes where it is required and its description. The “preferred source” column details the source for the reference material that was used for running the workflow.</p>

<table>
  <thead>
    <tr>
      <th>Input</th>
      <th>Required at</th>
      <th>Variable type</th>
      <th>File format</th>
      <th>Description</th>
      <th>Preferred source</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>is_paired_end</td>
      <td>All</td>
      <td>Boolean</td>
      <td>-</td>
      <td>Are input reads paired-end?</td>
      <td>-</td>
    </tr>
    <tr>
      <td>input_fastq_r1</td>
      <td>All</td>
      <td>Array[File]</td>
      <td>fastq</td>
      <td>Array of paths indicating the location of the fastq files to be processed. If paired-end, the path to the (_1) files.</td>
      <td>-</td>
    </tr>
    <tr>
      <td>input_fastq_r2</td>
      <td>All</td>
      <td>Array[File]</td>
      <td>fastq</td>
      <td>If paired-end, the path to the (_2) files. The position of each element in the array should match its pair in the input_fastq_r1 variable.</td>
      <td>-</td>
    </tr>
    <tr>
      <td>sample_id</td>
      <td>All</td>
      <td>Array[String]</td>
      <td>-</td>
      <td>Array of sample identifiers. Those identifiers will be used across the different tasks to the pipeline to identify each input sample. The position of each element in the array should match its pair in the input_fastq_r1 variable.</td>
      <td>-</td>
    </tr>
    <tr>
      <td>group</td>
      <td>All</td>
      <td>Array[String]</td>
      <td>-</td>
      <td>Array of string indicating the group of samples for each read file. Those groups will be used to perform the differential expression and signaling analyses. The position of each element in the array should match its pair in the input_fastq_r1 variable.</td>
      <td>-</td>
    </tr>
    <tr>
      <td>execution_mode</td>
      <td>-</td>
      <td>String</td>
      <td>-</td>
      <td>String indicating the execution mode to be used.</td>
      <td>-</td>
    </tr>
    <tr>
      <td>do_vc</td>
      <td>“salmon-star”, “salmon-hisat”, “hisat2”, “star”</td>
      <td>Boolean</td>
      <td>-</td>
      <td>Perform the variant calling? Only in case users do not want to extract and use variants from RNA-Seq data.</td>
      <td>-</td>
    </tr>
    <tr>
      <td>gtf_file</td>
      <td>All</td>
      <td>File</td>
      <td>gtf</td>
      <td>Annotation file for the genome used at alignment and variant calling. It is also the input to create the transcript-to-gene file neccesary for the salmon quantifications.</td>
      <td><a href="ftp://ftp.ensembl.org/pub/release-99/gtf/homo_sapiens/Homo_sapiens.GRCh38.99.gtf.gz">ENSEMBL</a></td>
    </tr>
    <tr>
      <td>ref_fasta</td>
      <td>“salmon-star”, “salmon-hisat”, “hisat2”, “star”</td>
      <td>File</td>
      <td>fasta</td>
      <td>Reference genome used to perform the variant calling.</td>
      <td><a href="ftp://ftp.ensembl.org/pub/release-99/fasta/homo_sapiens/dna/Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz">ENSEMBL</a></td>
    </tr>
    <tr>
      <td>ref_fasta_index</td>
      <td>“salmon-star”, “salmon-hisat”, “hisat2”, “star”</td>
      <td>File</td>
      <td>fai</td>
      <td>Reference genome index used to perform the variant calling.</td>
      <td>Created from ref_fasta with samtools</td>
    </tr>
    <tr>
      <td>ref_dict</td>
      <td>“salmon-star”, “salmon-hisat”, “hisat2”, “star”</td>
      <td>File</td>
      <td>dict</td>
      <td>Reference genome dictionary used to perform the variant calling.</td>
      <td>Created from ref_fasta with samtools</td>
    </tr>
    <tr>
      <td>db_snp_vcf</td>
      <td>“salmon-star”, “salmon-hisat”, “hisat2”, “star”</td>
      <td>File</td>
      <td>vcf</td>
      <td>Database of SNPs used to perform the variant calling.</td>
      <td><a href="https://ftp.ncbi.nih.gov/snp/organisms/human_9606_b150_GRCh38p7/VCF/All_20170710.vcf.gz">NCBI</a></td>
    </tr>
    <tr>
      <td>db_snp_vcf_index</td>
      <td>“salmon-star”, “salmon-hisat”, “hisat2”, “star”</td>
      <td>File</td>
      <td>tbi</td>
      <td>SNP database index.</td>
      <td><a href="https://ftp.ncbi.nih.gov/snp/organisms/human_9606_b150_GRCh38p7/VCF/00-common_all.vcf.gz.tbi">NCBI</a></td>
    </tr>
    <tr>
      <td>known_vcfs</td>
      <td>“salmon-star”, “salmon-hisat”, “hisat2”, “star”</td>
      <td>Array[File]</td>
      <td>vcf</td>
      <td>Databases of INDELs used to perform the variant calling.</td>
      <td><a href="https://ftp.ncbi.nih.gov/snp/organisms/human_9606_b150_GRCh38p7/VCF/All_20170710.vcf.gz">NCBI</a></td>
    </tr>
    <tr>
      <td>known_vcfs_indices</td>
      <td>“salmon-star”, “salmon-hisat”, “hisat2”, “star”</td>
      <td>Array[File]</td>
      <td>tbi</td>
      <td>INDEL databases indices.</td>
      <td><a href="https://ftp.ncbi.nih.gov/snp/organisms/human_9606_b150_GRCh38p7/VCF/00-common_all.vcf.gz.tbi">NCBI</a></td>
    </tr>
    <tr>
      <td>vep_cache_dir</td>
      <td>“salmon-star”, “salmon-hisat”, “hisat2”, “star”</td>
      <td>String</td>
      <td>-</td>
      <td>Path to the vep cache directory containing the variant annotations to be used. Should contain the SIFT and PolyPhen scores.</td>
      <td><a href="ftp://ftp.ensembl.org/pub/release-99/variation/indexed_vep_cache/">ENSEMBL</a></td>
    </tr>
    <tr>
      <td>hisat2_index_path</td>
      <td>“salmon-hisat”, “hisat2”</td>
      <td>String</td>
      <td>-</td>
      <td>Path to the directory where the HISAT2 index is stored.</td>
      <td>Created from ref_fasta with HISAT2</td>
    </tr>
    <tr>
      <td>hisat2_index_prefix</td>
      <td>“salmon-hisat”, “hisat2”</td>
      <td>String</td>
      <td>-</td>
      <td>HISAT2 index prefix.</td>
      <td>-</td>
    </tr>
    <tr>
      <td>star_index_path</td>
      <td>“salmon-star”, “star”</td>
      <td>String</td>
      <td>-</td>
      <td>Path to the directory where the STAR index is stored.</td>
      <td>Created from ref_fasta with STAR</td>
    </tr>
    <tr>
      <td>salmon_index_path</td>
      <td>“salmon”, “salmon-star”, “salmon-hisat2”</td>
      <td>String</td>
      <td>-</td>
      <td>Path to the directory where the Salmon index is stored.</td>
      <td>Created from ref_fasta with Salmon</td>
    </tr>
    <tr>
      <td>edger_script</td>
      <td>All</td>
      <td>File</td>
      <td>-</td>
      <td>Script executed in the EdgeR task.</td>
      <td>Included in MIGNON</td>
    </tr>
    <tr>
      <td>tximport_script</td>
      <td>“salmon”, “salmon-star”, “salmon-hisat2”</td>
      <td>File</td>
      <td>-</td>
      <td>Script executed in the TxImport task.</td>
      <td>Included in MIGNON</td>
    </tr>
    <tr>
      <td>hipathia_script</td>
      <td>All</td>
      <td>File</td>
      <td>-</td>
      <td>Script executed in the hiPathia task.</td>
      <td>Included in MIGNON</td>
    </tr>
    <tr>
      <td>ensemblTx_script</td>
      <td>“salmon”, “salmon-star”, “salmon-hisat2”</td>
      <td>File</td>
      <td>-</td>
      <td>Script executed in the ensembldb task. It transforms the provided GTF into a Tx2Gene file used by tximport.</td>
      <td>Included in MIGNON</td>
    </tr>
  </tbody>
</table>

<h2 id="execution-values">Execution values</h2>

<p>The following list contains the input parameters that can be used to control the number of CPUs and memory assigned to the container and the process where the task is executed. We will only list parameters that <strong>really have an impact on the task execution</strong>, that is, those parameters that are actually passed not only to the runtime of the WDL task, but also to the command that runs the operation.</p>

<h2 id="default-values">Default values</h2>

<h3 id="description">Description</h3>

<p>For all the above inputs, the <strong>cpu</strong> and <strong>mem</strong> parameters are directly passed to the container that executes each task. On the other hand, the <strong>cpu</strong> parameter is also passed to the argument that control the multi-threading on each tool. Before modifying any of this parameters, please check the parallelization section.</p>

<p>The <strong>haplotype_scatter_count</strong> input requires a special mention. As explained in the parallelization section, this parameter allows to apply the <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360035532012-Parallelism-Multithreading-Scatter-Gather">scatter and gather strategy</a> to parallelize the <strong>GATK HaplotypeCaller</strong> sub-task. This input will determine the number of chunks in which the reference genome is divide to perform the variant calling from aligned reads.</p>

<h2 id="other-inputs">Other inputs</h2>

<p>In this section, users can find inputs which are not required or used to control the workflow execution, but that can notably affect the final circuit signaling activity values.</p>

<h3 id="table">Table</h3>

<table>
  <thead>
    <tr>
      <th>Input</th>
      <th>Variable type</th>
      <th>Default value</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>fastp_windows_size</td>
      <td>Int</td>
      <td>4</td>
    </tr>
    <tr>
      <td>fastp_mean_quality</td>
      <td>Int</td>
      <td>15</td>
    </tr>
    <tr>
      <td>fastp_required_length</td>
      <td>Int</td>
      <td>20</td>
    </tr>
    <tr>
      <td>salmon_library_type</td>
      <td>String</td>
      <td>A</td>
    </tr>
    <tr>
      <td>tx2gene_file</td>
      <td>File</td>
      <td>-</td>
    </tr>
    <tr>
      <td>edger_min_counts</td>
      <td>Int</td>
      <td>15</td>
    </tr>
    <tr>
      <td>hipathia_normalize</td>
      <td>Boolean</td>
      <td>“true”</td>
    </tr>
    <tr>
      <td>hipathia_ko_factor</td>
      <td>Float</td>
      <td>0.01</td>
    </tr>
    <tr>
      <td>vep_sift_cutoff</td>
      <td>Float</td>
      <td>0.05</td>
    </tr>
    <tr>
      <td>vep_polyphen_cutoff</td>
      <td>Float</td>
      <td>0.95</td>
    </tr>
  </tbody>
</table>

<h3 id="description-1">Description</h3>

<ul>
  <li><strong>fastp_windows_size</strong>: Sliding windows size used by fastp to evaluate the quality of the sequences.</li>
  <li><strong>fastp_mean_quality</strong>: Required mean quality for a sequence to pass the fastp filter.</li>
  <li><strong>fastp_required_length</strong>: Reads shorter than this parameter will be discarded.</li>
  <li><strong>salmon_library_type</strong>: Library type for salmon quantification. Defaults to “A” (automatic detection).</li>
  <li><strong>tx2gene_file</strong>: Transcript to gene table. This parameter can be used to avoid the GTF to Tx2Gene task (ensembldb). This file must contain a two columns table indicating the mapping between the regions quantified by salmon and the features that will be used to perform the subsequent analyses (genes). This is specially important if using a GTF file that does not come from from ENSEMBL.</li>
  <li><strong>edger_min_counts</strong>: Minimum counts per gene. This parameter will be used to filter the count matrix before the differential expression analysis. By default, all those genes with less than 15 counts across all samples will be filtered.</li>
  <li><strong>hipathia_normalize</strong>: Normalize circuit pathway activity by length? HiPathia applies a signal propagation algorithm across the different receptor-effector circuits in the pathways to calculate the signaling circuit activity. The topology and diameter of such circuits influence the resulting circuit activity values, so this normalization is recommended.</li>
  <li><strong>hipathia_ko_factor</strong>: LoF variants knockdown factor. This parameter control the number by which the normalized expression values are multiplied when a loss of function (LoF) variant is detected for a gene/sample.</li>
  <li><strong>vep_sift_cutoff</strong>: Sift cutoff value. In combination with <strong>vep_polyphen_cutoff</strong>, it is used to filter variants that will be considered as deleterious (LoF).</li>
  <li><strong>vep_polyphen_cutoff</strong>: Polyphen cutoff value. In combination with <strong>vep_sift_cutoff</strong>, it is used to filter variants that will be considered as deleterious (LoF).</li>
</ul>

<h2 id="defaults">Defaults</h2>

<p>Apart from the inputs above described, there are other inputs which are required for the workflow to be executed. Those are set as defaults, and can be changed directly in the <a href="https://github.com/babelomics/MIGNON/blob/master/wdl/MIGNON.wdl">MIGNON wdl code</a> and the <a href="https://github.com/babelomics/MIGNON/blob/master/wdl/MIGNON_calling.wdl">MIGNON variant calling sub-workflow</a>.</p>

      </div>
    </main><footer class="site-footer h-card">
  <data class="u-url" href="/"></data>

  <div class="wrapper">

    <h2 class="footer-heading">MIGNON</h2>

    <div class="footer-col-wrapper">
      <div class="footer-col footer-col-1">
        <ul class="contact-list">
          <li class="p-name">Martin Garrido Rodriguez-Cordoba</li></ul>
      </div>

      <div class="footer-col footer-col-2"><ul class="social-media-list"><li><a href="https://github.com/babelomics"><svg class="svg-icon"><use xlink:href="/assets/minima-social-icons.svg#github"></use></svg> <span class="username">babelomics</span></a></li><li><a href="https://www.twitter.com/ClinicalBioinfo"><svg class="svg-icon"><use xlink:href="/assets/minima-social-icons.svg#twitter"></use></svg> <span class="username">ClinicalBioinfo</span></a></li></ul>
</div>

      <div class="footer-col footer-col-3">
        <p>Bioinformatic workflow for the mechanistic integrative analysis of RNA-Seq data</p>
      </div>
    </div>

  </div>

</footer>
</body>

</html>
