<!DOCTYPE html>
<html lang="en"><head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1"><!-- Begin Jekyll SEO tag v2.6.1 -->
<title>Output | MIGNON</title>
<meta name="generator" content="Jekyll v3.9.0" />
<meta property="og:title" content="Output" />
<meta name="author" content="Martin Garrido Rodriguez-Cordoba" />
<meta property="og:locale" content="en_US" />
<meta name="description" content="Bioinformatic workflow for the mechanistic integrative analysis of RNA-Seq data" />
<meta property="og:description" content="Bioinformatic workflow for the mechanistic integrative analysis of RNA-Seq data" />
<link rel="canonical" href="http://localhost:4000/3_output.html" />
<meta property="og:url" content="http://localhost:4000/3_output.html" />
<meta property="og:site_name" content="MIGNON" />
<script type="application/ld+json">
{"@type":"WebPage","headline":"Output","url":"http://localhost:4000/3_output.html","author":{"@type":"Person","name":"Martin Garrido Rodriguez-Cordoba"},"description":"Bioinformatic workflow for the mechanistic integrative analysis of RNA-Seq data","@context":"https://schema.org"}</script>
<!-- End Jekyll SEO tag -->
<link rel="stylesheet" href="/assets/main.css"><link type="application/atom+xml" rel="alternate" href="http://localhost:4000/feed.xml" title="MIGNON" /></head>
<body><header class="site-header" role="banner">

  <div class="wrapper"><a class="site-title" rel="author" href="/">MIGNON</a><nav class="site-nav">
        <input type="checkbox" id="nav-trigger" class="nav-trigger" />
        <label for="nav-trigger">
          <span class="menu-icon">
            <svg viewBox="0 0 18 15" width="18px" height="15px">
              <path d="M18,1.484c0,0.82-0.665,1.484-1.484,1.484H1.484C0.665,2.969,0,2.304,0,1.484l0,0C0,0.665,0.665,0,1.484,0 h15.032C17.335,0,18,0.665,18,1.484L18,1.484z M18,7.516C18,8.335,17.335,9,16.516,9H1.484C0.665,9,0,8.335,0,7.516l0,0 c0-0.82,0.665-1.484,1.484-1.484h15.032C17.335,6.031,18,6.696,18,7.516L18,7.516z M18,13.516C18,14.335,17.335,15,16.516,15H1.484 C0.665,15,0,14.335,0,13.516l0,0c0-0.82,0.665-1.483,1.484-1.483h15.032C17.335,12.031,18,12.695,18,13.516L18,13.516z"/>
            </svg>
          </span>
        </label>

        <div class="trigger"><a class="page-link" href="/1_installation.html">Installation</a><a class="page-link" href="/2_input.html">Input</a><a class="page-link" href="/3_output.html">Output</a><a class="page-link" href="/4_advanced.html">Advanced</a></div>
      </nav></div>
</header>
<main class="page-content" aria-label="Content">
      <div class="wrapper">
        <h1 id="output">Output</h1>

<p>As described in the <strong>tool manuscript</strong>, the main output of the workflow consists on two files, which can be found in the <strong>call-hipathia</strong> task directory created by <a href="https://github.com/broadinstitute/cromwell">cromwell</a>.</p>

<h2 id="circuit-activity-matrix">Circuit activity matrix</h2>

<p><strong>path_values.tsv</strong></p>

<p>This matrix contain the results of the circuit activity estimation performed by hiPathia after applying the <em>in-silico knockdown</em> strategy and contain a value for each circuit (row) and sample (column).</p>

<table>
  <thead>
    <tr>
      <th> </th>
      <th>ERR3481954</th>
      <th>ERR3481955</th>
      <th>ERR3481956</th>
      <th>ERR3481957</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>P-hsa03320-37</td>
      <td>0.0000</td>
      <td>0.0000</td>
      <td>0.0123</td>
      <td>0.0000</td>
    </tr>
    <tr>
      <td>P-hsa03320-61</td>
      <td>0.0320</td>
      <td>0.0286</td>
      <td>0.0123</td>
      <td>0.0485</td>
    </tr>
    <tr>
      <td>P-hsa03320-46</td>
      <td>0.0000</td>
      <td>0.0000</td>
      <td>0.0000</td>
      <td>0.0000</td>
    </tr>
    <tr>
      <td>P-hsa03320-57</td>
      <td>0.0000</td>
      <td>0.0000</td>
      <td>0.0123</td>
      <td>0.0000</td>
    </tr>
    <tr>
      <td>P-hsa03320-64</td>
      <td>0.0000</td>
      <td>0.0000</td>
      <td>0.0000</td>
      <td>0.0000</td>
    </tr>
    <tr>
      <td>P-hsa03320-47</td>
      <td>0.0808</td>
      <td>0.0768</td>
      <td>0.0301</td>
      <td>0.0693</td>
    </tr>
    <tr>
      <td>P-hsa03320-65</td>
      <td>0.2524</td>
      <td>0.2466</td>
      <td>0.2550</td>
      <td>0.2727</td>
    </tr>
    <tr>
      <td>P-hsa03320-55</td>
      <td>0.3917</td>
      <td>0.3999</td>
      <td>0.4006</td>
      <td>0.4027</td>
    </tr>
    <tr>
      <td>P-hsa03320-56</td>
      <td>0.2252</td>
      <td>0.2238</td>
      <td>0.2306</td>
      <td>0.2976</td>
    </tr>
  </tbody>
</table>

<h2 id="differential-signaling-activity-analysis">Differential signaling activity analysis</h2>

<p>This data frame contains the results of making all the possible pairwise comparisons between the circuit activity values of the sample groups specified by the <strong>group</strong> input. Those results are generated through the application of an iterative Wilcoxon Test. The group contrast is indicated in the “comparison” column.</p>

<p><strong>differential_signaling.tsv</strong></p>

<table>
  <thead>
    <tr>
      <th>pathId</th>
      <th>UP.DOWN</th>
      <th>statistic</th>
      <th>p.value</th>
      <th>FDRp.value</th>
      <th>comparison</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>P-hsa03320-37</td>
      <td>DOWN</td>
      <td>-1.5275</td>
      <td>0.2000</td>
      <td>0.7026</td>
      <td>16_weeks-12_weeks</td>
    </tr>
    <tr>
      <td>P-hsa03320-61</td>
      <td>DOWN</td>
      <td>-0.2182</td>
      <td>1.0000</td>
      <td>1.0000</td>
      <td>16_weeks-12_weeks</td>
    </tr>
    <tr>
      <td>P-hsa03320-46</td>
      <td>DOWN</td>
      <td>-1.9640</td>
      <td>0.1000</td>
      <td>0.6091</td>
      <td>16_weeks-12_weeks</td>
    </tr>
    <tr>
      <td>P-hsa03320-57</td>
      <td>DOWN</td>
      <td>-1.9640</td>
      <td>0.1000</td>
      <td>0.6091</td>
      <td>16_weeks-12_weeks</td>
    </tr>
    <tr>
      <td>P-hsa03320-64</td>
      <td>UP</td>
      <td>1.0911</td>
      <td>0.3537</td>
      <td>0.8566</td>
      <td>16_weeks-12_weeks</td>
    </tr>
    <tr>
      <td>P-hsa03320-47</td>
      <td>DOWN</td>
      <td>-1.0911</td>
      <td>0.4000</td>
      <td>0.8566</td>
      <td>16_weeks-12_weeks</td>
    </tr>
    <tr>
      <td>P-hsa03320-65</td>
      <td>UP</td>
      <td>0.2182</td>
      <td>1.0000</td>
      <td>1.0000</td>
      <td>16_weeks-12_weeks</td>
    </tr>
    <tr>
      <td>P-hsa03320-55</td>
      <td>UP</td>
      <td>0.6547</td>
      <td>0.7000</td>
      <td>0.9215</td>
      <td>16_weeks-12_weeks</td>
    </tr>
    <tr>
      <td>P-hsa03320-56</td>
      <td>DOWN</td>
      <td>-1.0911</td>
      <td>0.4000</td>
      <td>0.8566</td>
      <td>16_weeks-12_weeks</td>
    </tr>
    <tr>
      <td>…</td>
      <td>…</td>
      <td>…</td>
      <td>…</td>
      <td>…</td>
      <td>…</td>
    </tr>
    <tr>
      <td>P-hsa05321-94</td>
      <td>DOWN</td>
      <td>-1.3093</td>
      <td>0.1967</td>
      <td>0.5469</td>
      <td>9_weeks-16_weeks</td>
    </tr>
    <tr>
      <td>P-hsa05321-95</td>
      <td>UP</td>
      <td>1.9640</td>
      <td>0.1000</td>
      <td>0.4000</td>
      <td>9_weeks-16_weeks</td>
    </tr>
    <tr>
      <td>P-hsa05321-122</td>
      <td>UP</td>
      <td>1.9640</td>
      <td>0.0765</td>
      <td>0.4000</td>
      <td>9_weeks-16_weeks</td>
    </tr>
    <tr>
      <td>P-hsa05321-123</td>
      <td>UP</td>
      <td>0.2182</td>
      <td>1.0000</td>
      <td>1.0000</td>
      <td>9_weeks-16_weeks</td>
    </tr>
    <tr>
      <td>P-hsa05321-55</td>
      <td>UP</td>
      <td>1.9640</td>
      <td>0.1000</td>
      <td>0.4000</td>
      <td>9_weeks-16_weeks</td>
    </tr>
    <tr>
      <td>P-hsa05321-74</td>
      <td>UP</td>
      <td>1.3093</td>
      <td>0.1967</td>
      <td>0.5469</td>
      <td>9_weeks-16_weeks</td>
    </tr>
    <tr>
      <td>P-hsa05321-81</td>
      <td>UP</td>
      <td>1.3093</td>
      <td>0.1967</td>
      <td>0.5469</td>
      <td>9_weeks-16_weeks</td>
    </tr>
    <tr>
      <td>P-hsa05321-138</td>
      <td>DOWN</td>
      <td>-1.0911</td>
      <td>0.3537</td>
      <td>0.7966</td>
      <td>9_weeks-16_weeks</td>
    </tr>
    <tr>
      <td>P-hsa05321-75</td>
      <td>UP</td>
      <td>1.9640</td>
      <td>0.1000</td>
      <td>0.4000</td>
      <td>9_weeks-16_weeks</td>
    </tr>
    <tr>
      <td>P-hsa05321-152</td>
      <td>UP</td>
      <td>1.0911</td>
      <td>0.4000</td>
      <td>0.7966</td>
      <td>9_weeks-16_weeks</td>
    </tr>
  </tbody>
</table>

<h2 id="pathway-viewer">Pathway viewer</h2>

<p>The above results can be visualized using the <a href="https://bioconductor.org/packages/release/bioc/html/hipathia.html">hiPathia package</a>, by subsetting the table to the comparison of interest and using the <code class="language-plaintext highlighter-rouge">create_report()</code> function.</p>

<p><img src="https://github.com/babelomics/hipathia/blob/master/vignettes/pics/hipathia_report_1.png?raw=true" alt="Viewer" /></p>

<h2 id="other-outputs">Other outputs</h2>

<h3 id="knockdown-matrix">Knockdown matrix</h3>

<p><strong>ko_matrix.tsv</strong></p>

<p>This matrix contains the gene table that is used to perform the in-silico knockdown of genes that present a Loss of Function (LoF) variant and that therefore, can not propagate the signal across the circuits.</p>

<table>
  <thead>
    <tr>
      <th> </th>
      <th>ERR2704712</th>
      <th>ERR2704713</th>
      <th>ERR2704714</th>
      <th>ERR2704715</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>ENSG00000128342</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
    </tr>
    <tr>
      <td>ENSG00000187860</td>
      <td>0.01</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
    </tr>
    <tr>
      <td>ENSG00000100003</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
    </tr>
    <tr>
      <td>ENSG00000100012</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
    </tr>
    <tr>
      <td>ENSG00000181123</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
    </tr>
    <tr>
      <td>ENSG00000133488</td>
      <td>1</td>
      <td>1</td>
      <td>0.01</td>
      <td>1</td>
    </tr>
    <tr>
      <td>ENSG00000128242</td>
      <td>1</td>
      <td>0.01</td>
      <td>1</td>
      <td>1</td>
    </tr>
    <tr>
      <td>ENSG00000100029</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
    </tr>
    <tr>
      <td>ENSG00000185339</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
      <td>1</td>
    </tr>
  </tbody>
</table>

<h3 id="inner-tools-outputs">Inner tools outputs</h3>

<p>Apart from the outputs derived directly from the MIGNON strategy, some of the tools employed during the raw reads processing generate intermediate output reports which can be explored in its corresponding call directories. Additionally, we strongly recommend to use the <a href="https://multiqc.info/">MultiQC</a> tool to summarize all those intermediate outputs in a single html report.</p>

      </div>
    </main><footer class="site-footer h-card">
  <data class="u-url" href="/"></data>

  <div class="wrapper">

    <h2 class="footer-heading">MIGNON</h2>

    <div class="footer-col-wrapper">
      <div class="footer-col footer-col-1">
        <ul class="contact-list">
          <li class="p-name">Martin Garrido Rodriguez-Cordoba</li></ul>
      </div>

      <div class="footer-col footer-col-2"><ul class="social-media-list"><li><a href="https://github.com/babelomics"><svg class="svg-icon"><use xlink:href="/assets/minima-social-icons.svg#github"></use></svg> <span class="username">babelomics</span></a></li><li><a href="https://www.twitter.com/ClinicalBioinfo"><svg class="svg-icon"><use xlink:href="/assets/minima-social-icons.svg#twitter"></use></svg> <span class="username">ClinicalBioinfo</span></a></li></ul>
</div>

      <div class="footer-col footer-col-3">
        <p>Bioinformatic workflow for the mechanistic integrative analysis of RNA-Seq data</p>
      </div>
    </div>

  </div>

</footer>
</body>

</html>
