# MIGNON

MIGNON is a bioinformatic workflow for the mechanistic integrative analysis of rna-seq data. It is written using the [Workflow Description Language (WDL)](https://github.com/openwdl/wdl) and can be executed using [Cromwell](https://github.com/broadinstitute/cromwell). It implements a novel way of analyzing RNA-Seq data, extracting the transcriptomic and genomic information obtainable from RNA-Seq reads. By using a *in-silico* knockdown strategy, it estimates the cellular signaling circuits activities through the application of the [hipathia](http://hipathia.babelomics.org/) model, using gene expression as a proxy of protein signaling activity.

## Installation

In order to execute the workflow, the user needs to be able of launching the [Cromwell](https://github.com/broadinstitute/cromwell) engine, regardless of where it is deployed. We recommend reading the [Five Minutes Intro](https://cromwell.readthedocs.io/en/stable/tutorials/FiveMinuteIntro/) created by the Broad Institute team, which explain the basic operations that can be performed using their software. Additionally, as the tools employed by the workflow are used as [docker](https://www.docker.com/) containers, the system where the pipeline is deployed should have an engine to run such containers. Particularly, we have tested the workflow both locally with [Docker](https://www.docker.com/) and in a High Performance Computing (HPC) environment, executing the containers as [Singularity](https://sylabs.io/guides/3.5/user-guide/) images. Theorically, it is also deployable within the [Terra](https://terra.bio/) platform, which makes use of cloud computing services to execute such workflows and analyze and visualize outputs.
