# MIGNON TUTORIAL

MIGNON is a bioinformatic workflow for the mechanistic integrative analysis of rna-seq data. It is written using the [Workflow Description Language (WDL)](https://github.com/openwdl/wdl) and can be executed using [Cromwell](https://github.com/broadinstitute/cromwell). It implements a novel way of analyzing RNA-Seq data, extracting the transcriptomic and genomic information obtainable from RNA-Seq reads. By using a *in-silico* knockdown strategy, it estimates the cellular signaling circuits activities through the application of the [hipathia](http://hipathia.babelomics.org/) model, using gene expression as a proxy of protein signaling activity.

## Installation

In order to execute the workflow, the user needs to have the required software to launch the [Cromwell](https://github.com/broadinstitute/cromwell) engine. We recommend reading the [Five Minutes Intro](https://cromwell.readthedocs.io/en/stable/tutorials/FiveMinuteIntro/) created by the Broad Institute team to their software, as they give a walkthrough to the basic operations that can be performed using this tool. Additionally, as the tools employed by the workflow are used as [docker](https://www.docker.com/) containers, the system where the pipeline is deployed should have an engine to run such containers. Particularly, we have tested the workflow both locally with [docker](https://www.docker.com/) and in a High Performance Computing (HPC) environment, executing the containers as [Singularity](https://sylabs.io/guides/3.5/user-guide/) images.

## Execution modes

Regarding the execution of the workflow, as explained in the **tool manuscript**, we have designed 5 modes of execution that make use of different tools in crucial steps of the workflow. 

| Execution mode  | Description                                                                                                                                                                                      |
|-----------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| "salmon-hisat2" | Alignment with HISAT2. Count table obtained from Salmon quantifications. Deals with multi-mapping reads. Allows variant calling. Low memory consumption during alignment, but slower than STAR.  |
| "salmon-star"   | Alignment with STAR. Count table obtained from Salmon quantifications. Deals with multi-mapping reads. Allows variant calling. High memory consumption during alignment, but faster than HISAT2. |
| "hisat2"        | Alignment with HISAT2. Count table obtained with featureCounts. Does not deal with multi-mapping reads. Allows variant calling. Low memory consumption during alignment, but slower than STAR.   |
| "star"          | Alignment with STAR. Count table obtained with featureCounts. Does not deal with multi-mapping reads. Allows variant calling. High memory consumption during alignment, but faster than HISAT2.  |
| "salmon"        | Count table obtained from Salmon quantifications. Deals with multi-mapping reads. Does not allow variant calling.                                                                                |

We strongly recommend to use the combined execution strategies “salmon-star” or “salmon-hisat2”, as they use the pseudo-alignment strategy to quantify gene expression dealing with multi-mapping reads, and star or hisat2 to obtain the alignments for the variant calling sub-workflow.

## Preparing the input

As explained by [WDL authors](https://github.com/openwdl/wdl/blob/master/versions/development/SPEC.md#specifying-workflow-inputs-in-json), cromwell uses a [JSON](https://www.json.org/) formatted file as input. We have prepared a list of JSON files with the minimum required inputs for each execution mode. They can be found at the [input_templates](https://github.com/babelomics/MIGNON/input_templates/) folder.

## Containers

The following table contains the list of containers used during the execution of the workflow:

| Tool                       | Version                                         | Docker container                                                                                                                                                                                        | File Format | Description                                                                                                                                                                                                                                                  | Preferred Source                                                                                                      |
|----------------------------|-------------------------------------------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|-------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| fastp                      | v0.20.0                                         | quay.io/biocontainers/fastp:0.20.0                                                                                                                                                                      | -           | Are input reads paired-end?                                                                                                                                                                                                                                  | -                                                                                                                     |
| fastqc                     | v0.11.5                                         | [biocontainers/fastqc:v0.11.5_cv4](https://hub.docker.com/layers/biocontainers/fastqc/v0.11.5_cv4/images/sha256-387748462c7fc280b7959ceda0f6251190d2e4b9ebc0585d24e7bcb58bdcf2bf?context=explore)       | fastq       | Array of paths indicating the location of the fastq files to be processed. If paired-end, the path to the (_1) files.                                                                                                                                        | -                                                                                                                     |
| samtools                   | v1.9                                            | quay.io/biocontainers/samtools:1.9                                                                                                                                                                      | fastq       | If paired-end, the path to the (_2) files. The position of each element in the array should match its pair in the input_fastq_r1 variable.                                                                                                                   | -                                                                                                                     |
| HISAT2                     | v2.1.0                                          | quay.io/biocontainers/hisat2:2.1.0                                                                                                                                                                      | -           | Array of sample identifiers. Those identifiers will be used across the different tasks to the pipeline to identify each input sample. The position of each element in the array should match its pair in the input_fastq_r1 variable.                        | -                                                                                                                     |
| STAR                       | v.2.7.2b                                        | quay.io/biocontainers/star:2.7.2b                                                                                                                                                                       | -           | Array of string indicating the group of samples for each read file. Those groups will be used to perform the differential expression and signaling analyses. The position of each element in the array should match its pair in the input_fastq_r1 variable. | -                                                                                                                     |
| salmon                     | v.0.13.0                                        | quay.io/biocontainers/salmon:0.13.0                                                                                                                                                                     | -           | String indicating the execution mode to be used.                                                                                                                                                                                                             | -                                                                                                                     |
| GATK                       | v4.1.3.0                                        | [broadinstitute/gatk:4.1.3.0](https://hub.docker.com/layers/broadinstitute/gatk/4.1.3.0/images/sha256-e37193b61536cf21a2e1bcbdb71eac3d50dcb4917f4d7362b09f8d07e7c2ae50?context=explore)                 | -           | Perform the variant calling? Only in case users do not want to extract and use variants from RNA-Seq data.                                                                                                                                                   | -                                                                                                                     |
| picard                     | v2.20.7                                         | [broadinstitute/picard:2.20.7](https://hub.docker.com/layers/broadinstitute/picard/2.20.7/images/sha256-a8aee5af2e485b23c2498b6e9271133ab355a1e5e3c62a7e2b96f84ba60978ee?context=explore)               | gtf         | Annotation file for the genome used at alignment and variant calling. It is also the input to create the transcript-to-gene file neccesary for the salmon quantifications.                                                                                   | [ENSEMBL](ftp://ftp.ensembl.org/pub/release-99/gtf/homo_sapiens/Homo_sapiens.GRCh38.99.gtf.gz)                        |
| VeP                        | v99                                             | [ensemblorg/ensembl-vep:release_99.1](https://hub.docker.com/layers/ensemblorg/ensembl-vep/release_99.1/images/sha256-ca890d3d06d8ebddfb6126a1e4e257aa516f0522e75513994e797d97dca7c9af?context=explore) | fasta       | Reference genome used to perform the variant calling.                                                                                                                                                                                                        | [ENSEMBL](ftp://ftp.ensembl.org/pub/release-99/fasta/homo_sapiens/dna/Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz) |
| txImport                   | v1.10.0                                         | quay.io/biocontainers/bioconductor-tximport:1.10.0                                                                                                                                                      | fai         | Reference genome index used to perform the variant calling.                                                                                                                                                                                                  | Created from ref_fasta with samtools.                                                                                 |
| edgeR                      | v3.28.0                                         | quay.io/biocontainers/bioconductor-edger:3.28.0                                                                                                                                                         | dict        | Reference genome dictionary used to perform the variant calling.                                                                                                                                                                                             | Created from ref_fasta with samtools.                                                                                 |
| hipathia                   | v2.2.0                                          | quay.io/biocontainers/bioconductor-hipathia:2.2.0                                                                                                                                                       | vcf         | Database of SNPs used to perform the variant calling.                                                                                                                                                                                                        | [NCBI](https://ftp.ncbi.nih.gov/snp/organisms/human_9606_b150_GRCh38p7/VCF/All_20170710.vcf.gz)                       |
| MIGNON.db_snp_vcf_index    | "salmon-star", "salmon-hisat", "hisat2", "star" | File                                                                                                                                                                                                    | tbi         | SNP database index.                                                                                                                                                                                                                                          | [NCBI](https://ftp.ncbi.nih.gov/snp/organisms/human_9606_b150_GRCh38p7/VCF/00-common_all.vcf.gz.tbi)                  |
| MIGNON.known_vcfs          | "salmon-star", "salmon-hisat", "hisat2", "star" | Array[File]                                                                                                                                                                                             | vcf         | Databases of INDELs used to perform the variant calling.                                                                                                                                                                                                     | [NCBI](https://ftp.ncbi.nih.gov/snp/organisms/human_9606_b150_GRCh38p7/VCF/All_20170710.vcf.gz)                       |
| MIGNON.known_vcfs_indices  | "salmon-star", "salmon-hisat", "hisat2", "star" | Array[File]                                                                                                                                                                                             | tbi         | INDEL databases indices.                                                                                                                                                                                                                                     | [NCBI](https://ftp.ncbi.nih.gov/snp/organisms/human_9606_b150_GRCh38p7/VCF/00-common_all.vcf.gz.tbi)                  |
| MIGNON.vep_cache_dir       | "salmon-star", "salmon-hisat", "hisat2", "star" | String                                                                                                                                                                                                  | -           | Path to the vep cache directory containing the variant annotations to be used. Should contain the SIFT and PolyPhen scores.                                                                                                                                  | [ENSEMBL](ftp://ftp.ensembl.org/pub/release-99/variation/indexed_vep_cache/)                                          |
| MIGNON.hisat2_index_path   | "salmon-hisat", "hisat2"                        | String                                                                                                                                                                                                  | -           | Path to the directory where the HISAT2 index is stored.                                                                                                                                                                                                      | Created from ref_fasta with HISAT2.                                                                                   |
| MIGNON.hisat2_index_prefix | "salmon-hisat", "hisat2"                        | String                                                                                                                                                                                                  | -           | HISAT2 index prefix.                                                                                                                                                                                                                                         | -                                                                                                                     |
| MIGNON.star_index_path     | "salmon-star", "star"                           | String                                                                                                                                                                                                  | -           | Path to the directory where the STAR index is stored.                                                                                                                                                                                                        | Created from ref_fasta with STAR.                                                                                     |
| MIGNON.salmon_index_path   | "salmon", "salmon-star", "salmon-hisat2"        | String                                                                                                                                                                                                  | -           | Path to the directory where the Salmon index is stored.                                                                                                                                                                                                      | Created from ref_fasta with STAR.                                                                                     |
| MIGNON.edger_script        | All                                             | File                                                                                                                                                                                                    | -           | Script executed in the EdgeR task.                                                                                                                                                                                                                           | Included in MIGNON.                                                                                                   |
| MIGNON.tximport_script     | "salmon", "salmon-star", "salmon-hisat2"        | File                                                                                                                                                                                                    | -           | Script executed in the TxImport task.                                                                                                                                                                                                                        | Included in MIGNON.                                                                                                   |
| MIGNON.hipathia_script     | All                                             | File                                                                                                                                                                                                    | -           | Script executed in the hiPathia task.                                                                                                                                                                                                                        | Included in MIGNON.                                                                                                   |
| MIGNON.ensemblTx_script    | "salmon", "salmon-star", "salmon-hisat2"        | File                                                                                                                                                                                                    | -           | Script executed in the ensembldb task. It transforms the provided GTF into a Tx2Gene file used by tximport.                                                                                                                                                  | Included in MIGNON.                                                                                                   |                                                                                      |

## Inputs

### Required inputs

| Input                      | Required at                                     | Variable Type | File Format | Description                                                                                                                                                                                                                                                  | Preferred Source                                                                                                      |
|----------------------------|-------------------------------------------------|---------------|-------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| MIGNON.is_paired_end       | All                                             | Boolean       | -           | Are input reads paired-end?                                                                                                                                                                                                                                  | -                                                                                                                     |
| MIGNON.input_fastq_r1      | All                                             | Array[File]   | fastq       | Array of paths indicating the location of the fastq files to be processed. If paired-end, the path to the (_1) files.                                                                                                                                        | -                                                                                                                     |
| MIGNON.input_fastq_r2      | All                                             | Array[File]   | fastq       | If paired-end, the path to the (_2) files. The position of each element in the array should match its pair in the input_fastq_r1 variable.                                                                                                                   | -                                                                                                                     |
| MIGNON.sample_id           | All                                             | Array[String] | -           | Array of sample identifiers. Those identifiers will be used across the different tasks to the pipeline to identify each input sample. The position of each element in the array should match its pair in the input_fastq_r1 variable.                        | -                                                                                                                     |
| MIGNON.group               | All                                             | Array[String] | -           | Array of string indicating the group of samples for each read file. Those groups will be used to perform the differential expression and signaling analyses. The position of each element in the array should match its pair in the input_fastq_r1 variable. | -                                                                                                                     |
| MIGNON.execution_mode      | -                                               | String        | -           | String indicating the execution mode to be used.                                                                                                                                                                                                             | -                                                                                                                     |
| MIGNON.do_vc               | "salmon-star", "salmon-hisat", "hisat2", "star" | Boolean       | -           | Perform the variant calling? Only in case users do not want to extract and use variants from RNA-Seq data.                                                                                                                                                   | -                                                                                                                     |
| MIGNON.gtf_file            | All                                             | File          | gtf         | Annotation file for the genome used at alignment and variant calling. It is also the input to create the transcript-to-gene file neccesary for the salmon quantifications.                                                                                   | [ENSEMBL](ftp://ftp.ensembl.org/pub/release-99/gtf/homo_sapiens/Homo_sapiens.GRCh38.99.gtf.gz)                        |
| MIGNON.ref_fasta           | "salmon-star", "salmon-hisat", "hisat2", "star" | File          | fasta       | Reference genome used to perform the variant calling.                                                                                                                                                                                                        | [ENSEMBL](ftp://ftp.ensembl.org/pub/release-99/fasta/homo_sapiens/dna/Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz) |
| MIGNON.ref_fasta_index     | "salmon-star", "salmon-hisat", "hisat2", "star" | File          | fai         | Reference genome index used to perform the variant calling.                                                                                                                                                                                                  | Created from ref_fasta with samtools.                                                                                 |
| MIGNON.ref_dict            | "salmon-star", "salmon-hisat", "hisat2", "star" | File          | dict        | Reference genome dictionary used to perform the variant calling.                                                                                                                                                                                             | Created from ref_fasta with samtools.                                                                                 |
| MIGNON.db_snp_vcf          | "salmon-star", "salmon-hisat", "hisat2", "star" | File          | vcf         | Database of SNPs used to perform the variant calling.                                                                                                                                                                                                        | [NCBI](https://ftp.ncbi.nih.gov/snp/organisms/human_9606_b150_GRCh38p7/VCF/All_20170710.vcf.gz)                       |
| MIGNON.db_snp_vcf_index    | "salmon-star", "salmon-hisat", "hisat2", "star" | File          | tbi         | SNP database index.                                                                                                                                                                                                                                          | [NCBI](https://ftp.ncbi.nih.gov/snp/organisms/human_9606_b150_GRCh38p7/VCF/00-common_all.vcf.gz.tbi)                  |
| MIGNON.known_vcfs          | "salmon-star", "salmon-hisat", "hisat2", "star" | Array[File]   | vcf         | Databases of INDELs used to perform the variant calling.                                                                                                                                                                                                     | [NCBI](https://ftp.ncbi.nih.gov/snp/organisms/human_9606_b150_GRCh38p7/VCF/All_20170710.vcf.gz)                       |
| MIGNON.known_vcfs_indices  | "salmon-star", "salmon-hisat", "hisat2", "star" | Array[File]   | tbi         | INDEL databases indices.                                                                                                                                                                                                                                     | [NCBI](https://ftp.ncbi.nih.gov/snp/organisms/human_9606_b150_GRCh38p7/VCF/00-common_all.vcf.gz.tbi)                  |
| MIGNON.vep_cache_dir       | "salmon-star", "salmon-hisat", "hisat2", "star" | String        | -           | Path to the vep cache directory containing the variant annotations to be used. Should contain the SIFT and PolyPhen scores.                                                                                                                                  | [ENSEMBL](ftp://ftp.ensembl.org/pub/release-99/variation/indexed_vep_cache/)                                          |
| MIGNON.hisat2_index_path   | "salmon-hisat", "hisat2"                        | String        | -           | Path to the directory where the HISAT2 index is stored.                                                                                                                                                                                                      | Created from ref_fasta with HISAT2.                                                                                   |
| MIGNON.hisat2_index_prefix | "salmon-hisat", "hisat2"                        | String        | -           | HISAT2 index prefix.                                                                                                                                                                                                                                         | -                                                                                                                     |
| MIGNON.star_index_path     | "salmon-star", "star"                           | String        | -           | Path to the directory where the STAR index is stored.                                                                                                                                                                                                        | Created from ref_fasta with STAR.                                                                                     |
| MIGNON.salmon_index_path   | "salmon", "salmon-star", "salmon-hisat2"        | String        | -           | Path to the directory where the Salmon index is stored.                                                                                                                                                                                                      | Created from ref_fasta with STAR.                                                                                     |
| MIGNON.edger_script        | All                                             | File          | -           | Script executed in the EdgeR task.                                                                                                                                                                                                                           | Included in MIGNON.                                                                                                   |
| MIGNON.tximport_script     | "salmon", "salmon-star", "salmon-hisat2"        | File          | -           | Script executed in the TxImport task.                                                                                                                                                                                                                        | Included in MIGNON.                                                                                                   |
| MIGNON.hipathia_script     | All                                             | File          | -           | Script executed in the hiPathia task.                                                                                                                                                                                                                        | Included in MIGNON.                                                                                                   |
| MIGNON.ensemblTx_script    | "salmon", "salmon-star", "salmon-hisat2"        | File          | -           | Script executed in the ensembldb task. It transforms the provided GTF into a Tx2Gene file used by tximport.                                                                                                                                                  | Included in MIGNON.                                                                                                   |